makefile kernel improvements

dannf opened the pull request:
- Avoid using predictable paths in `/tmp`
- Split up the target into multiple chunks to ease debugging
- Add kernel dep to targets that call melange build when using qemu runner
- Fix `QEMU_KERNEL_IMAGE` being set in melange's environment
- Cleanup kernel w/ `clean` target

dannf committed:
Makefile: put .PHONY annotation next to target

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Makefile: fetch-kernel improvements

- Avoid using predictable paths in /tmp
- Split up the target into multiple chunks to ease debugging
- Add kernel dep to targets that call melange build when using qemu runner
- Fix QEMU_KERNEL_IMAGE being set in melange's environment
- Cleanup kernel w/ clean target

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

mattmoor commented on Makefile:
```diff
@@ -98,22 +105,43 @@ ${CACHEDIR}/.libraries_token.txt: cache
-	export QEMU_KERNEL_IMAGE=/tmp/kernel/boot/vmlinuz
-	export MELANGE_OPTS="--runner=qemu"
+.PHONY: fetch-kernel
+fetch-kernel:
+	rm -rf kernel
+	$(MAKE) kernel/boot/vmlinuz
+
+kernel/APKINDEX.tar.gz:
+	$(MAKE) apk-token
+	mkdir -p kernel
+	curl -LS --silent -o $@ \
+	  --user user:$$(chainctl auth token --audience apk.cgr.dev) \
+	  https://apk.cgr.dev/chainguard-private/$(ARCH)/APKINDEX.tar.gz
+
+kernel/APKINDEX: kernel/APKINDEX.tar.gz
+	tar -x -C kernel -f $< $(notdir $@)
+	touch $@
+
+kernel/chosen: kernel/APKINDEX
+	grep -e '^P:' -e '^V:' $< | sed -n '/^P:linux$$/{n;p}' > \
+	  kernel/available
+	grep '^V:' kernel/available | sed 's/V://' | \
+	  sort -V | tail -n1 > kernel/chosen
+
+kernel/linux.apk: kernel/chosen
+	$(MAKE) apk-token
+	kver=$(file < kernel/chosen); \
+	  curl -LS --silent -o $@.tmp \
+	  --user user:$$(chainctl auth token --audience apk.cgr.dev) \
+	  https://apk.cgr.dev/chainguard-private/$(ARCH)/linux-$$kver.apk
+	mv $@.tmp $@
+
+kernel/boot/vmlinuz: kernel/linux.apk
+	tar -x -C kernel -f $< boot/ 2> /dev/null
+	touch $@
```
JFYI I copied the original block from enterprise, so let's update there as well!

mattmoor commented on Makefile:
```diff
@@ -19,6 +19,12 @@ SOURCES ?= gs://wolfi-sources/
 ifneq (${MELANGE_RUNNER},)
 	MELANGE_OPTS += --runner ${MELANGE_RUNNER}
 endif
+QEMU_KERNEL_IMAGE ?= kernel/boot/vmlinuz
+ifeq (${MELANGE_RUNNER},qemu)
+	QEMU_KERNEL_DEP = ${QEMU_KERNEL_IMAGE}
+	export QEMU_KERNEL_IMAGE
+endif
```
Does this magic make it so that we fetch the kernel when the melange runner is set to QEMU?  Amazing 🤩

mattmoor reviewed (commented):
I'm a Makefile noob, so going to defer to @smoser here, but this LGTM!

dannf committed:
Makefile: Use tmpfile for curl consistently

^C'ing a curl download will leave a partial file in place. Make doesn't
know it is partial - it only goes by the timestamp. So let's download
to a temporary file and only rename to the target filename once it is
complete.

We do this with the linux.apk download already. Do it also for the
index for consistency.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Makefile: add and use authget function

From Scott Moser.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

smoser commented on Makefile:
```diff
@@ -98,22 +105,44 @@ ${CACHEDIR}/.libraries_token.txt: cache
+	grep -e '^P:' -e '^V:' $< | sed -n '/^P:linux$$/{n;p}' > \
+	  kernel/available
+	grep '^V:' kernel/available | sed 's/V://' | \
+	  sort -V | tail -n1 > kernel/chosen
```
2 things here. when you redirect like this, you create the file and then it will be up to date even if it fails.
it'd be good to verify that there is content you expected and then move it into place.

```
out=$$(grep '^V:' kernel/available | sed 's/V://' | sort -V | tail -n1) && [ -n "$$out" ] && echo "$$out" > $@)

smoser replied:
something like that, then you it doesnt get updated unless its sane.

dannf committed:
Makefile: Sanity check chosen kernel version

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

smoser reviewed (approved):
love it. thanks.
